src/model_signing/manifest.py

```diff
@@ -245,16 +246,156 @@ def key(self) -> ManifestKey:
         return Shard(self._path, self._start, self._end)
 
 
+class SerializationType(metaclass=abc.ABCMeta):
+    """A description of the serialization process that generated the manifest.
+
+    These should record all the parameters needed to ensure a reproducible
+    serialization. These are used to build a manifest back from the signature in
+    a backwards compatible way. We use these to determine what serialization to
+    use when verifying a signature.
+    """
+
+    @property
+    @abc.abstractmethod
+    def serialization_dict(self) -> dict[str, Any]:
+        """The arguments of the serialization method."""
+
+    @classmethod
+    def from_dict(cls, serialization_dict: dict[str, Any]) -> Self:
+        """Builds an instance of this class based on the dict representation.
+
+        This is the reverse of `serialization_dict`.
+        """
+        serialization_type = serialization_dict["method"]
+        for subclass in [_FileSerialization, _ShardSerialization]:
+            if serialization_type == subclass.method:
+                return subclass._from_dict(serialization_dict)
+        raise ValueError(f"Unknown serialization type {serialization_type}")
+
+    @classmethod
+    @abc.abstractmethod
+    def _from_dict(cls, serialization_dict: dict[str, Any]) -> Self:
+        """Performs the actual build from `from_dict`."""
+
+    @abc.abstractmethod
+    def new_item(self, name: str, digest: hashing.Digest) -> ManifestItem:
+        """Returns a new manifest item with given name and digest."""
+
+
+class _FileSerialization(SerializationType):
+    method: Final[str] = "files"
+
+    def __init__(self, hash_type: str, allow_symlinks: bool = False):
```

Comment:
does it matter that these are positional?

Reply:
They cannot be confused since they have different types, but we could add an `*` and make them keyword based if really needed.